dttools/src/copy_tree.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2013- The University of Notre Dame
+ * This software is distributed under the GNU General Public License.
+ * See the file COPYING for details.
+ */
+
+#include <assert.h>
+#include <dirent.h>
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <string.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "copy_tree.h"
+#include "copy_stream.h"
+#include "create_dir.h"
+#include "debug.h"
+#include "path.h"
+#include "xxmalloc.h"
+
+mode_t default_dirmode = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
+
+int copy_dir_real(const char *source, const char *target);
+
+int copy_symlink(const char *source, const char *target) {
+	char linkname[PATH_MAX];
+	ssize_t r;
+
+	assert(source);
+	assert(target);
+
+	if(!access(target, F_OK)) {
+		LDEBUG("%s already exists!\n", target);
+		return -1;
+	}
+
+	r = readlink(source, linkname, sizeof(linkname));
+	if(r == -1) {
+		LDEBUG("readlink(`%s`) failed: %s!\n", source, strerror(errno));
+		return -1;
+	}
+	linkname[r] = '\0';
+
+	if(symlink(linkname, target)) {
+		LDEBUG("symlink(`%s`, `%s`) failed: %s\n", linkname, target, strerror(errno));
+		return -1;
+	}
+	return 0;
+}
+
+/* return 0 on success, return -1 on failure. */
+int copy_direntry(const char *s, const char *t) {
+	struct stat s_stat;
+
+	assert(s);
+	assert(t);
+
+	if(lstat(s, &s_stat)) {
+		LDEBUG("lstat(`%s`): %s\n", s, strerror(errno));
+		return -1;
+	}
+
+	if(S_ISDIR(s_stat.st_mode)) {
+		if(!create_dir(t, (int)default_dirmode)) return -1;
+		return copy_dir_real(s, t);
+	} else if(S_ISREG(s_stat.st_mode)) {
+		return copy_file_to_file(s, t);
+	} else if(S_ISLNK(s_stat.st_mode)) {
+		return copy_symlink(s, t);
+	} else {
+		LDEBUG("Ignore Copying %s: only dir, regular files, and symlink are supported!\n", s);
```

Comment:
I meant: "Not copying...".  "Ignore copying" sounds as if you are copying but ignoring the results.